[AJ-1460] Add notice to dashboard if workspace has group constraint policies

nawatts opened the pull request:
https://broadworkbench.atlassian.net/browse/AJ-1460

This adds a notice to the workspace dashboard if the workspace has group constraint policies.

Clicking on the info icon shows a message describing data access controls and listing the groups that a user must be a member of in order to gain access to the workspace.

<img width="453" alt="Screenshot 2023-11-14 at 7 26 46 PM" src="https://github.com/DataBiosphere/terra-ui/assets/1156625/f7271917-124a-4bdb-b119-33cc193b4bb2">

<img width="313" alt="Screenshot 2023-11-14 at 7 26 52 PM" src="https://github.com/DataBiosphere/terra-ui/assets/1156625/b3ad5c37-1bc9-4398-b778-2edaf8743f53">

## Testing

To mock group constraint policies for a workspace, run the following the console (replacing NAMESPACE and NAME with the namespace and name of your workspace) and click the "Dashboard" tab to refresh it.

```js
window.ajaxOverridesStore.set([
  {
    filter: { url: /api\/workspaces\/NAMESPACE\/NAME/ },
    fn: window.ajaxOverrideUtils.mapJsonBody(workspace => {
      return {
        ...workspace,
        policies: [
          {
            namespace: 'terra',
            name: 'group-constraint',
            additionalData: [{ group: 'example-group' }]
          }
        ]
      }
    })
  }
]);
```

nawatts committed:
Add notice to dashboard if workspace has group constraint policies

nawatts commented on src/pages/workspaces/workspace/Dashboard/WorkspaceInformation.test.ts:
```diff
@@ -99,4 +101,42 @@ describe('WorkspaceInformation', () => {
+        {
+          namespace: 'terra',
+          name: 'group-constraint',
+          additionalData: [{ group: 'foo' }, { group: 'bar' }],
```
I don't know if a group constraint policy can actually have multiple groups listed in `additionalData`.

cahrens replied:
Can you ask folks on your team to see if we should expect multiple "group-constraint" policies or just a single one with multiple groups?

davidangb replied:
It can be both. There can be multiple group-constraint policies, and each of those policies can have multiple groups underneath it.

aherbst-broad replied:
@davidangb almost. WSM will balk if multiple group constraint policies are submitted. The request may have multiple groups in the `additionalData` field, as is expressed in the code above, or this example:
```
 "policies": [
    {
      "name": "group-constraint",
      "namespace": "terra",
      "additionalData": [
        {
          "group": "arh-testing-group"
        },
        {
          "group": "monster-dev"
        }
      ]
    }
```

Submitting a creation request with the above payload gives me a workspace with an auth domain that looks like this:
![Screenshot 2023-11-15 at 11 15 50 AM](https://github.com/DataBiosphere/terra-ui/assets/74209955/c51753cf-0d56-450d-b450-a859b373b3b2)

cahrens replied:
Ah, thanks @aherbst-broad! I have no memory of us having that Authorization Domain section. It seems like we should only show one or the other… not both.

nawatts replied:
It sounds like Rawls returns group constraint policies as auth domains then? Should we be using the `authorizationDomain` field instead of `policies` then? Or do we need to add an `if (isGcpWorkspace)` condition to every use of the `authorizationDomain` field in Terra UI?

nawatts replied:
Also, maybe we should use a similar design to what we have for showing auth domains for group constraint policies? It has a lot more room for the list of groups.

cahrens replied:
It looks like it doesn't appear in policies at all….

```
 {
        "accessLevel": "NO ACCESS",
        "public": false,
        "workspace": {
            "workspaceId": "b4e2fc4d-b8a1-4f04-8758-94ccd62bc4ea",
            "name": "arhPolicyTesting9",
            "state": "Ready",
            "lastModified": "2023-11-15T16:09:24.442Z",
            "attributes": {},
            "cloudPlatform": "Azure",
            "createdBy": "aherbst@broadinstitute.org",
            "namespace": "dsp-dev-testing-20230921",
            "authorizationDomain": [
                {
                    "membersGroupName": "monster-dev"
                },
                {
                    "membersGroupName": "arh-testing-group"
                }
            ]
        }
    },

nawatts replied:
Does it if you include `policies` in the `fields` parameters to the list workspaces API? I don't think that field is returned by default.

cahrens replied:
Yes, it does. Drew said it's expected that it will be returned in both fields.


![image](https://github.com/DataBiosphere/terra-ui/assets/484484/6ba9548b-41c0-44e8-ab38-cee63904dbad)

cahrens commented on src/libs/workspace-utils.ts:
```diff
@@ -164,6 +164,30 @@ export const regionConstraintMessage = (workspace: BaseWorkspace): string | unde
+    _.uniq,
+    _.sortBy(_.identity)
+  )(groupConstraintPolicies);
+  return `Data Access Controls add additional permission restrictions to a workspace. These were added when you imported data from a controlled access source. All workspace collaborators must also be members of the following groups: ${requiredGroups.join(
```
I would suggest using pluralize for "group(s)".

nawatts replied:
Done in https://github.com/DataBiosphere/terra-ui/pull/4469/commits/8fae28195d036c4bb1953b4c15151806bba1d93e.

cahrens commented on src/pages/workspaces/workspace/Dashboard/WorkspaceInformation.test.ts:
```diff
@@ -99,4 +101,43 @@ describe('WorkspaceInformation', () => {
+        {
+          namespace: 'terra',
+          name: 'group-constraint',
+          additionalData: [{ group: 'foo' }],
```
Until we know that multiple `group-constraint` policies are not possible, I would suggest you change this so that the group name is different, and then make sure that your message shows all 3 group names.

nawatts replied:
I reused a group name in both policies because I wanted to test that the message de-duplicated group names.

cahrens replied:
Then can you add a second, different group in this one? In addition to testing that it dedupes, it would be nice to know that it actually pulls the group names from both (instead of only respecting the second group policy, for example).

nawatts replied:
Ooh, I see what you mean. Yes.

nawatts replied:
I removed the second group here and added a test for `groupConstraintMessage` to test de-duplication in https://github.com/DataBiosphere/terra-ui/pull/4469/commits/34ed68d9e54ee0ff8b0af44e9e46325c2eda55a3.

nawatts committed:
Pluralize "groups" in message

cahrens commented on src/libs/workspace-utils.ts:
```diff
@@ -183,9 +184,10 @@ export const groupConstraintMessage = (workspace: WorkspaceWrapper): string | un
-    ', '
-  )}.`;
+  return `Data Access Controls add additional permission restrictions to a workspace. These were added when you imported data from a controlled access source. All workspace collaborators must also be members of the following ${pluralize(
+    'groups',
```
Does this work? I've always passed in the singular for it to pluralize.

nawatts replied:
Turns out it does:
```
> const pluralize = (await import('pluralize')).default;
undefined
> pluralize('groups', 1)
'group'
> pluralize('groups', 3)
'groups'
```

But this was not intentional. Changed to match convention in https://github.com/DataBiosphere/terra-ui/pull/4469/commits/e65a25037b01478f64dd37b5fea8bcc92e19e9ee.

nawatts committed:
Pass singular form to pluralize

nawatts committed:
Update tests

davidangb commented on src/libs/workspace-utils.test.ts:
```diff
@@ -158,6 +161,102 @@ describe('hasRegionConstraint', () => {
+        },
+      ],
+      expectedResult: true,
+    },
```
suggested test case: the workspace has a protected-data policy but no group-constraint policies; it should still return false:
```suggestion
    },
    {
      policies: [
        {
          namespace: 'terra',
          name: 'protected-data',
          additionalData: [],
        },
      ],
      expectedResult: false,
    },
```

nawatts replied:
Added in https://github.com/DataBiosphere/terra-ui/pull/4469/commits/ef57e8e5e21382d5b59ce6cf6d6ae189e3793139.

davidangb reviewed (commented):
One suggestion on test cases inline (which may apply to multiple unit tests).

Re: The actual message of "must also be members of the following groups: bar, foo" … I thought that since the group names can be confusing, Product wanted to hide those for now until we can ensure the names we output are helpful. Please verify this verbiage with Product/UX?

nawatts commented:
> I thought that since the group names can be confusing, Product wanted to hide those for now until we can ensure the names we output are helpful. 

Yeah, I took some initiative with this text. After talking to @cahrens yesterday, we thought the messages about data access controls were pretty useless without information about how a collaborator could meet them / which groups they needed to join.

nawatts commented on src/libs/workspace-utils.ts:
```diff
@@ -164,6 +165,31 @@ export const regionConstraintMessage = (workspace: BaseWorkspace): string | unde
+    _.uniq,
+    _.sortBy(_.identity)
+  )(groupConstraintPolicies);
+  return `Data Access Controls add additional permission restrictions to a workspace. These were added when you imported data from a controlled access source. All workspace collaborators must also be members of the following ${pluralize(
+    'group',
+    requiredGroups.length
+  )}: ${requiredGroups.join(', ')}.`;
```
@bethsheets here's the PR with the text for the group constraint / data access controls notice on the workspace dashboard.

bethsheets replied:
@nawatts @cahrens I know it is not ideal, but we could not agree on what the policy name should be (AnVIL data access troubleshooting is complex) and so it was moved to post-MVP because our first priority is making sure the end to end flow works. Identi-team is leading the initiative to identify the right data access control terms and what a troubleshooting journey looks like.  We don't have AnVIL users on Azure quite yet, so we have time to make this better. We agreed on being generic for the time being "data access controls are present"

nawatts replied:
I'm confused... it sounds like we've decided on a term: "data access controls". What prevents being able to show the groups required?

nawatts committed:
Add test case

nawatts committed:
Rename functions and update text

nawatts commented:
Updated the message text to remove the list of groups. Also renamed the function to more closely match the name used in the UI vs the policy data model.

cahrens commented:
@nawatts don't you think we should also remove the Authorization Domain section for Azure workspaces?

nawatts commented:
Yes. I filed a separate ticket for it ([AJ-1473](https://broadworkbench.atlassian.net/browse/AJ-1473)), but I can add that here if preferred.

davidangb reviewed (approved):
thanks for all the followup on this one!
